src/test/scala/chiseltest/tests/BackPressureTest.scala

```diff
@@ -0,0 +1,138 @@
+import chiseltest.experimental.TestOptionBuilder.ChiselScalatestOptionBuilder
+import chiseltest.internal.WriteVcdAnnotation
+import chiseltest._
+import org.scalatest.freespec.AnyFreeSpec
+
+import scala.util.Random
+
+class GcdInputBundle(val w: Int) extends Bundle {
+  val value1 = UInt(w.W)
+  val value2 = UInt(w.W)
+}
+
+class GcdOutputBundle(val w: Int) extends Bundle {
+  val value1 = UInt(w.W)
+  val value2 = UInt(w.W)
+  val gcd    = UInt(w.W)
+}
+/**
+  * Compute GCD using subtraction method.
+  * Subtracts the smaller of registers x and y from the larger until register y is zero.
+  * value input register x is then the Gcd
+  * returns a packet of information with the two input values and their GCD
+  */
+class DecoupledGcd(width: Int) extends Module {
+
+  val input = IO(Flipped(Decoupled(new GcdInputBundle(width))))
+  val output = IO(Decoupled(new GcdOutputBundle(width)))
+
+  val xInitial    = Reg(UInt())
+  val yInitial    = Reg(UInt())
+  val x           = Reg(UInt())
+  val y           = Reg(UInt())
+  val busy        = RegInit(false.B)
+
+  input.ready := ! busy
+  output.bits := DontCare
+  output.valid := false.B
+
+  when(busy)  {
+    // during computation keep subtracting the smaller from the larger
+    when(y > 0.U) {
+      when(x > y) {
+        x := x - y
+      }.otherwise {
+        y := y - x
+      }
+    }.otherwise {
+      // when y becomes zero computation is over, signal valid data to output
+      output.bits.value1 := xInitial
+      output.bits.value2 := yInitial
+      output.bits.gcd := x
+      output.valid := true.B
+      busy := ! output.ready
+    }
+  }.otherwise {
+    when(input.valid) {
+      // valid data available and no computation in progress, grab new values and start
+      val bundle = input.deq()
+      x := bundle.value1
+      y := bundle.value2
+      xInitial := bundle.value1
+      yInitial := bundle.value2
+      busy := true.B
+      output.valid := false.B
+    }
+  }
+}
+
+
+class BackPressureTest extends AnyFreeSpec with ChiselScalatestTester {
+  "This test fails with fork issues" in {
+
+    test(new DecoupledGcd(16)).withAnnotations(Seq(WriteVcdAnnotation)) { dut =>
+      dut.input.initSource().setSourceClock(dut.clock)
+      dut.output.initSink().setSinkClock(dut.clock) // the sink uses a slower clock so that backpressure is more likely
+
+      val testValues = for {x <- 1 to 10; y <- 1 to 10} yield (x, y)
+      val inputSeq = testValues.map { case (x, y) =>
+        (new GcdInputBundle(16)).Lit(_.value1 -> x.U, _.value2 -> y.U)
+      }
+      val resultSeq = testValues.map { case (x, y) =>
+        new GcdOutputBundle(16).Lit(_.value1 -> x.U, _.value2 -> y.U, _.gcd -> BigInt(x).gcd(BigInt(y)).U)
+      }
+
+      fork {
+        dut.input.enqueueSeq(inputSeq)
+        println("all data enqueued")
+      }.fork {
+        for(expected <- resultSeq) {
+          dut.output.expectDequeue(expected)
+          dut.output.ready.poke(false.B)
```

Comment:
I think the issue is the poke here, you have a poke without a gating step right after the inner timescope ends. It's weird?

In any case, the explicit poke should also be unnecessary, because with `initSink`, ready defaults to false when not otherwise driven (for example, in a `expectDequeue`) - one of the design goals of chiseltest was to avoid state errors from needing to manually revert signals, so the `timescope` construct handles it for you.